File: packages/sdk-maui/ScreebUtils.cs
namespace Screeb.Maui;

public static class ScreebUtils
{
    /// <summary>
    /// Converts DateTime/DateTimeOffset values to ISO 8601 strings with timezone offset,
    /// e.g. "2024-01-15T10:30:00.000+02:00". Recursively processes nested dictionaries.
    /// Mirrors Flutter's _formatDates and React Native's normalizeValue.
    /// </summary>
    public static Dictionary<string, object>? FormatProperties(Dictionary<string, object>? props)
    {
        if (props == null) return null;
        var result = new Dictionary<string, object>(props.Count);
        foreach (var (key, value) in props)
            result[key] = FormatValue(value);
        return result;
    }

    private static object FormatValue(object value) => value switch
    {
        DateTimeOffset dto => FormatDateTimeOffset(dto),
        DateTime dt => FormatDateTimeOffset(new DateTimeOffset(dt)),
        Dictionary<string, object> nested => FormatProperties(nested)!,
        _ => value
    };

    private static string FormatDateTimeOffset(DateTimeOffset dto)
    {
        var offset = dto.Offset;
        var sign = offset >= TimeSpan.Zero ? "+" : "-";
        var absHours = Math.Abs(offset.Hours).ToString("D2");
        var mins = Math.Abs(offset.Minutes).ToString("D2");
        return $"{dto:yyyy-MM-ddTHH:mm:ss.fff}{sign}{absHours}:{mins}";
    }
}

File: packages/sdk-kmp/src/commonMain/kotlin/app/screeb/sdk/kmp/ScreebUtils.kt
package app.screeb.sdk.kmp

/**
 * Converts Map<String, Any>? to a form safe for native SDK consumption.
 * Nested maps are recursively processed.
 * Mirrors MAUI's ScreebUtils.FormatProperties.
 */
internal object ScreebUtils {
    fun formatProperties(props: Map<String, Any>?): Map<String, Any>? {
        if (props == null) return null
        return props.mapValues { (_, v) -> formatValue(v) }
    }

    private fun formatValue(value: Any): Any = when (value) {
        is Map<*, *> -> {
            @Suppress("UNCHECKED_CAST")
            formatProperties(value as Map<String, Any>) ?: emptyMap<String, Any>()
        }
        else -> value
    }
}
